api/client/push.go

```diff
@@ -42,6 +50,136 @@ func (cli *DockerCli) CmdPush(args ...string) error {
 	v := url.Values{}
 	v.Set("tag", tag)
 
-	_, _, err = cli.clientRequestAttemptLogin("POST", "/images/"+remote+"/push?"+v.Encode(), nil, cli.out, repoInfo.Index, "push")
-	return err
+	streamOut := cli.out
+	targets := []target{}
+	trustedPush := isTrusted(remote)
+	if trustedPush {
+		// Create multiWriter
+		r, w := io.Pipe()
+		defer w.Close()
+		// Capture output
+		streamOut = io.MultiWriter(streamOut, w)
+
+		go func() {
+			scanner := bufio.NewScanner(r)
+			scanner.Split(ScanInput)
+			// New splitter
+			for scanner.Scan() {
+				line := scanner.Text()
+				if matches := targetRegexp.FindStringSubmatch(line); len(matches) == 4 {
+					dgst, err := digest.ParseDigest(matches[2])
+					if err != nil {
+						fmt.Fprintf(cli.out, "Bad digest value %q, ignoring\n", matches[2])
+						continue
+					}
+					s, err := strconv.ParseInt(matches[3], 10, 64)
+					if err != nil {
+						fmt.Fprintf(cli.out, "Bad size value %q, ignoring\n", matches[3])
+						continue
+					}
+
+					targets = append(targets, target{
+						reference: registry.ParseReference(matches[1]),
+						digest:    dgst,
+						size:      s,
+					})
+				}
+			}
+
+		}()
+	}
+
+	_, _, err = cli.clientRequestAttemptLogin("POST", "/images/"+remote+"/push?"+v.Encode(), nil, streamOut, repoInfo.Index, "push")
+	if err != nil {
+		return err
+	}
+	if !trustedPush {
+		return nil
+	}
+
+	if tag == "" {
+		fmt.Fprintf(cli.out, "No tag specified, skipping TUF metadata push\n")
+		return nil
+	}
+	if len(targets) == 0 {
+		fmt.Fprintf(cli.out, "No targets found, skipping TUF metadata push\n")
+		return nil
+	}
+
+	fmt.Fprintf(cli.out, "Signing and pushing TUF metadata")
+
+	canonicalName := registry.CanonicalizeName(remote)
+	if remote != canonicalName {
+		fmt.Fprintf(cli.out, "Expanded name to %s\n", canonicalName)
+	}
+
+	repo, err := cli.getNotaryRepository(canonicalName)
+	if err != nil {
+		fmt.Fprintf(cli.out, "Error establishing connection to notary repository: %s\n", err)
+		return err
+	}
+
+	for _, target := range targets {
+		hex, err := hex.DecodeString(target.digest.Hex())
+		if err != nil {
+			return err
+		}
+		t := &client.Target{
+			Name: target.reference.String(),
+			Hashes: data.Hashes{
+				string(target.digest.Algorithm()): hex,
+			},
+			Length: int64(target.size),
+		}
+		if err := repo.AddTarget(t); err != nil {
+			return err
+		}
+	}
+
+	passFunc := func() (string, error) {
+		passphrase, err := cli.getPassphrase()
+		if err != nil {
+			return "", err
+		}
+		return passphrase, nil
+	}
+
+	err = repo.Publish(passFunc)
+	if _, ok := err.(*client.ErrRepoNotInitialized); ok {
+		keys := repo.ListRootKeys()
+		var rootKey string
+
+		if len(keys) == 0 {
+			fmt.Fprintf(cli.out, "Creating a new private key\n")
+		} else {
+			rootKey = keys[0]
+			fmt.Fprintf(cli.out, "Using private key %q\n", rootKey)
+		}
+		passphrase, err := passFunc()
+		if err != nil {
+			return err
+		}
+		if rootKey == "" {
```

Comment:
i just feel like some of these ifs can be avoided with a different implementation